client/rpc/src/lib.rs

```diff
@@ -417,6 +446,86 @@ where
         Ok(forest_proof.encode())
     }
 
+    async fn generate_proof(
+        &self,
+        provider_id: H256,
+        seed: H256,
+        challenged_file_keys: Vec<H256>,
+    ) -> RpcResult<Vec<u8>> {
+        // TODO: Get provider ID itself.
+        // TODO: Generate challenges from seed.
+
+        // Generate the Forest proof in a closure to drop the read lock on the Forest Storage.
+        let proven_file_keys = {
+            // The Forest Key is an empty vector since this is a BSP, therefore it doesn't
+            // have multiple Forest keys.
+            let fs = self
+                .forest_storage_handler
+                .get(&Vec::new().into())
+                .await
+                .ok_or_else(|| {
+                    into_rpc_error(format!(
+                        "Forest storage not found for empty key. Make sure you're running a BSP."
+                    ))
+                })?;
+
+            let p = fs
+                .read()
+                .await
+                .generate_proof(challenged_file_keys.clone())
+                .map_err(into_rpc_error)?;
+
+            p
+        };
+
+        // Get the keys that were proven.
+        let mut proven_keys = Vec::new();
+        for key in proven_file_keys.proven {
+            match key {
+                Proven::ExactKey(leaf) => proven_keys.push(leaf.key),
+                Proven::NeighbourKeys((left, right)) => match (left, right) {
+                    (Some(left), Some(right)) => {
+                        proven_keys.push(left.key);
+                        proven_keys.push(right.key);
+                    }
+                    (Some(left), None) => proven_keys.push(left.key),
+                    (None, Some(right)) => proven_keys.push(right.key),
+                    (None, None) => {
+                        return Err(into_rpc_error("Both left and right leaves in forest proof are None. This should not be possible."));
+                    }
+                },
+                Proven::Empty => {
+                    return Err(into_rpc_error("Forest proof generated with empty forest. This should not be possible, as this provider shouldn't have been challenged with an empty forest."));
```

Comment:
This is an RPC method, right? So it might be called without the need of a challenge which means the error message might not be correct. I'd leave it as "Tried to generate forest proof with an empty forest" or something along those lines

Reply:
Yeah, I agree. I'll change that in the next PR to speed up this one getting merged.